server/gfmd/user.c

```diff
@@ -156,39 +157,68 @@ grpassign_add_group(struct group_assignment *ga)
 	u->groups.group_prev = ga;
 }
 
+static gfarm_error_t
+user_enter_auth_id(
+	enum auth_user_id_type auth_user_id_type,
+	char *auth_user_id, struct user *u);
+
 gfarm_error_t
-user_auth_id_modify(struct user *u, const char *auth_id_type,
-	const char *auth_user_id)
+user_auth_id_modify(struct user *u, char *auth_id_type,
+	char *auth_user_id)
```

Comment:
The following changes are needed.

* rename user_auth_id_modify() to user_auth_id_modify_internal() or something like that.
* add "bool *need_to_update_dbp" and "bool *need_to_addp" as the last parameters of user_auth_id_modify_internal().
  these are output parameters.
* check strcmp(u->auth_user_id[auth_user_id_type], auth_user_id) == 0;
  if these strings are equal, do "*need_to_update_dbp = false;" and return GFARM_ERR_NO_ERROR;
  (Note that without this check, the GFARM_ERR_ALREDY_EXISTS case below does not work correctly).
* lookup user_id_map to make sure the new auth_user_id is NOT already registered.
  If the new auth_user_id is already registered, this function should fail with GFARM_ERR_ALREDY_EXISTS.
  See GFARM_ERR_ALREDY_EXISTS case in user_enter_gsi_dn().
* use strdup_log() instead of strdup(), and return GFARM_ERR_NO_MEMORY if strdup_log() returns NULL. (note that EVERY ERROR should be checked, SIGSEGV is not allowed for this kind of server programs).
* check the return value of user_enter_auth_id(), and if it fails, revert the change to u->auth_user_id[auth_user_id_type].
* if old value of u->auth_user_id[auth_user_id_type] was NULL,
  then do "*need_to_addp = true",
  otherwise do "*need_to_addp = false";
  do "*need_to_update_dbp = true;" in both cases.
* add a new function user_auth_id_modify() which internally calls user_auth_id_modify_internal().
  if user_auth_id_modify_internal() fails, call gflog_fatal() with appropriate log messages,
  e.g.
  if the error was GFARM_ERR_INVALID_ARGUMENT, then gflog_fatal(, "user %s: unknown auth_id_type %s, slave gfmd is older than gfmd", auth_id_type).
  or if the error was GFARM_ERR_ALREDY_EXISTS, then gflog_fatal(, "user %s: unexpected inconsistency, adding duplicate auth_user_id_type:%s, auth_user_id:%s")